clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,380 @@ package common
-			if err != nil {
-				errList.Append(fmt.Errorf("reading object name error: %v", err))
-				return
-			}
-			objNamespace, err := runtimeobjects.GetNamespaceFromRuntimeObject(runtimeObjectsList[objectIndex])
-			if err != nil {
-				errList.Append(fmt.Errorf("reading object namespace error: %v", err))
-				return
-			}
-			glog.Infof("%s: %s (%s) has all pods running", waitForControlledPodsRunningName, objName, objNamespace)
-		})
+func (w *waitForControlledPodsRunningMeasurement) getObjectCountAndMaxVersion(c clientset.Interface) (int, uint64, error) {
+	var desiredCount int
+	var maxResourceVersion uint64
+	objects, err := runtimeobjects.ListRuntimeObjectsForKind(c, w.kind, w.namespace, w.labelSelector, w.fieldSelector)
+	if err != nil {
+		return desiredCount, maxResourceVersion, fmt.Errorf("listing objects error: %v", err)
 	}
-	wg.Wait()
 
-	if int(runningRuntimeObjects) == len(runtimeObjectsList) {
-		glog.Infof("%s: %d / %d %ss are running with all pods", waitForControlledPodsRunningName, runningRuntimeObjects, len(runtimeObjectsList), kind)
-		return summaries, nil
+	for i := range objects {
+		runtimeObj, ok := objects[i].(runtime.Object)
+		if !ok {
+			glog.Infof("%s: cannot cast to runtime.Object: %v", objects[i])
+			continue
+		}
+		version, err := runtimeobjects.GetResourceVersionFromRuntimeObject(runtimeObj)
+		if err != nil {
+			glog.Errorf("%s: retriving resource version error: %v", waitForControlledPodsRunningName, err)
+			continue
+		}
+		desiredCount++
+		if version > maxResourceVersion {
+			maxResourceVersion = version
+		}
 	}
-	return summaries, fmt.Errorf("error while waiting for %ss to be running in namespace '%v' with labels '%v' and fields '%v' - only %d found running with all pods: %v",
-		kind, namespace, labelSelector, fieldSelector, runningRuntimeObjects, errList.String())
+	return desiredCount, maxResourceVersion, nil
 }
 
-func waitForRuntimeObject(c clientset.Interface, obj runtime.Object, timeout time.Duration) error {
+func (w *waitForControlledPodsRunningMeasurement) waitForRuntimeObject(clientSet clientset.Interface, obj runtime.Object) (*objectChecker, error) {
 	runtimeObjectNamespace, err := runtimeobjects.GetNamespaceFromRuntimeObject(obj)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	runtimeObjectSelector, err := runtimeobjects.GetSelectorFromRuntimeObject(obj)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	runtimeObjectReplicas, err := runtimeobjects.GetReplicasFromRuntimeObject(obj)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	stopCh := make(chan struct{})
-	time.AfterFunc(timeout, func() {
-		close(stopCh)
+	o := newObjectCheker()
+	o.lock.Lock()
+	defer o.lock.Unlock()
+	w.handlingGroup.Start(func() {
+		err = waitForPods(clientSet, runtimeObjectNamespace, runtimeObjectSelector.String(), "", int(runtimeObjectReplicas), o.stopCh, false)
+		o.lock.Lock()
+		defer o.lock.Unlock()
+		key, keyErr := runtimeobjects.CreateMetaNamespaceKey(obj)
```

Comment:
compute it outside lock (or probably even in the outer function)

Reply:
Done.

Reply:
Why not compute it in the outer function?
if the object is incorrect, we have a way to return an error there.